### PR TITLE
Eng 11977 net addr str

### PR DESCRIPTION
**Pull request by BillWhite**
This adds four new functions to translate internet addresses, both IPv4 and IPv6, to and from presentation format and binary.

https://issues.voltdb.com/browse/ENG-11977

### CODE REVIEW

File: tests/ee/expressions/function_test.cpp [hunk 1]
```diff
@@ -551,6 +557,15 @@ TEST_F(FunctionTest, NaturalModTest) {
     ASSERT_EQ(sawException, true);
 }
 
+TEST_F(FunctionTest, inet6NtoATest) {
+    in6_addr addr;
```
Comment (BillWhite) [inline tests/ee/expressions/function_test.cpp]: Most of the tests are in the jUnit.  This is the only function that needed a C++ unit test to debug a mistake, and I decided to leave it in.

File: tests/frontend/org/voltdb/regressionsuites/TestFunctionsForVoltDBSuite.java [hunk 2]
```diff
@@ -2355,7 +2379,289 @@ public void testBin() throws NoConnectionsException, IOException, ProcCallExcept
...
+                                 (short)0x0000,
+                                 (short)0x0000,
+                                 (short)0x0000,
+                                 (short)0x0000,
+                                 (short)0x0000,
+                                 (short)0x0001
+                         });
+    }
+
+    public void notestDateadd() throws NoConnectionsException, IOException, ProcCallException {
```
Comment (wolffcm) [inline tests/frontend/org/voltdb/regressionsuites/TestFunctionsForVoltDBSuite.java]: I don't see any negative tests for converting invalid presentations to the binary representation.  Is it possible to get errors going the other way too (binary to presentation)?  We should definitely have negative tests in before you merge.

File: tests/frontend/org/voltdb/regressionsuites/TestFunctionsForVoltDBSuite.java [hunk 3]
```diff
@@ -287,7 +311,7 @@ static private void loadProcedures(VoltProjectBuilder project) {
         project.addProcedures(BadParamTypesForTimestamp.class);
     }
 
-    public void testExplicitErrorUDF() throws Exception
+    public void notestExplicitErrorUDF() throws Exception
```
Comment (BillWhite) [inline tests/frontend/org/voltdb/regressionsuites/TestFunctionsForVoltDBSuite.java]: Oops.  Sorry.

File: tests/frontend/org/voltdb/regressionsuites/TestFunctionsForVoltDBSuite.java (hunk 2 above)
Comment (BillWhite) [inline tests/frontend/org/voltdb/regressionsuites/TestFunctionsForVoltDBSuite.java]: I've added some for the aton functions. 

It's essentially impossible for us to fail when converting binary to presentation.  The functions which do the conversion are inet_ntop and inet6_ntop.  They can fail if their (binary) input buffer is too small, if the address family is not a known value or if there is not enough output buffer space to hold the output strings.  But we check for the first, and the second and third are compile time constants.  I think any 128-bit long sequence is a legal address for something.  Some have special interpretations, but they are plausible in this context.

File: src/ee/expressions/bitwisefunctions.h [hunk 4]
```diff
@@ -231,5 +232,158 @@ template<> inline NValue NValue::call<FUNC_VOLT_BIT_SHIFT_RIGHT>(const std::vect
...
+    // in the network byte order.  So we translate it
+    // using htonl.
+    addr.s_addr = htonl(v & mask32);
+    // Save space for the address string,
+    // plus one more for a null terminator.
+    char answer[INET_ADDRSTRLEN + 1];
+    answer[sizeof(answer)-1] = 0;
+    if (inet_ntop(AF_INET, &addr, answer, sizeof(answer)) == 0) {
+        throw SQLException(SQLException::dynamic_sql_error,
+                           errno,
```
Comment (BillWhite) [inline src/ee/expressions/bitwisefunctions.h]: I sent the error message to Andrew.  I'll see what he has to say.

File: src/ee/expressions/bitwisefunctions.h [hunk 5]
```diff
@@ -231,5 +232,158 @@ template<> inline NValue NValue::call<FUNC_VOLT_BIT_SHIFT_RIGHT>(const std::vect
...
+    if (INET6_ADDRSTRLEN < addrlen) {
+        std::stringstream sb;
+        sb << "INET6_ATON: Argument string is too long to be an IPv6 address: "
+           << addrlen;
+        throw SQLException(SQLException::dynamic_sql_error, sb.str().c_str());
+    }
+    // Copy the string out to cbuff so we can
+    // null terminate it for inet_pton.
+    char cbuff[INET6_ADDRSTRLEN + 1];
+    memcpy(cbuff, addr_str, addrlen);
```
Comment (BillWhite) [inline src/ee/expressions/bitwisefunctions.h]: I was bummed about that as well.

**Pull request merged**